manual/configuration.md
&#9665; [Installation](installation.md)
&nbsp;&nbsp;&nbsp;&nbsp; &#8801; [Table of Contents](README.md)
&nbsp;&nbsp;&nbsp;&nbsp; [Starting and Stopping](starting-and-stopping.md) &#9655;
- - -

# Configuration
Seanox Devwex uses a central preconfigured configuration `devwex.ini` which is
located in the current working directory `./devwex/program`. The configuration
is divided into sections with keys and values, which also supports dynamic
values and access to system and environment variables.


## Content
- [Configuration File](#configuration-file)
- [Configuration Structure](#configuration-structure)
- [General Application Configuration](#general-application-configuration)
- [Initialisation](#initialisation)
- [Server](#server)
- [Remote Control Server](#remote-control-server)
- [HTTP / HTTPS Server](#http--https-server)
  - [Server](#server-1)
  - [Transport Layer Security / Secure Socket Layer](#transport-layer-security--secure-socket-layer)
  - [Client Authentication / Mutual Authentication](#client-authentication--mutual-authentication)
  - [Virtual Hosts / Virtual Hosting](#virtual-hosts--virtual-hosting)
  - [Filters](#filters)
  - [Virtual Paths (Aliasing and Redirection)](#virtual-paths-aliasing-and-redirection)
  - [Basic Access Authentication / Digest Access Authentication](#basic-access-authentication--digest-access-authentication)
  - [Directory Listing](#directory-listing)
  - [Environment Variables](#environmental-variables)
  - [Common Gateway Interface](#common-gateway-interface)
- [Status Codes](#status-codes)
- [Media Types](#media-types)


## Configuration File
The configuration uses an extended INI format which, like the original format,
is line-based and uses sections with keys with values, whereby sections and keys
are case-insensitive and declarations with the same name cause sections and keys
to be overwritten. One extension is that sections can use multiple inheritance
by using the keyword `EXTENDS` followed by the names of the referenced sections.
In this way, all keys and values of the referenced sections are adopted as the
basis and can then be overwritten in the section.

The values are assigned to the keys with an equals sign. This assignment can be
continued in the next line if this starts with a plus sign. Values can also be
fixed, variable and optionally assigned. With the option `[?]` at the end of a
key, the system searches for a key with the same name in the system properties
of the Java runtime environment. If no suitable value is found, an optionally
specified value is used. Without a value, these keys are interpreted as
non-existent and ignored.


Comments begin with a semicolon or, more commonly, a hash. They are optional and
can appear anywhere in a line; all characters following them are then no longer
part of the section, key, or value. The `[+]` option at the end of a key can be
used to disable the comment function for that key, allowing a semicolon or hash
in the value.

> [!NOTE]
> __The semicolon (`;`) is the original comment character for this format. The
> hash (`#`) is also supported, which is common in many modern configuration
> formats. Both characters have the same function. For simplicity, the original
> semicolon is used throughout the following documentation; however, the hash
> sign can be used interchangeably.__

Comments begin with a semicolon, are optional and can be used anywhere in a
line, so the following characters are not part of section, key or value. With
the option `[+]` at the end of a key, the use of comments for that key can be
disabled and the semicolon used in the value.

Sections, keys and values also support a hexadecimal notation, starting with
`0x...` followed by the hexadecimal string, which can only be used for the
complete element.

For better readability, spaces can be inserted before the keys, which will be
ignored by the configuration.

> __Example__
> ```
> [SECTION] EXTENDS SECTION-A SECTION-B      ;comment
>   PARAM-A                 = VALUE-1        ;comment
>   PARAM-B             [+] = VALUE-2; VALUE-3
>                           + VALUE-4; VALUE-5
>   PARAM-C          [?][+] = VALUE-6; VALUE-7
>   PARAM-E          [?]                     ;comment
>
> [0x53454354494F4E2D41]
>   PARAM-A                 = 0x574552542D31 ;comment
>   0x504152414D2D42        = VALUE-2        ;comment
>   0x504152414D2D43    [+] = 0x574552542D33
>   PARAM-D                 = 0x574552542D34
>                           + 0x574552542D35
>   PARAM-E          [?][+] = 0x574552542D363B20574552542D37
>   0x504152414D2D45 [?]                     ;comment
> ```

__The indentation has no relevance, it is only intended to improve
readability.__

> __Line 1__
> ```
> [SECTION] EXTENDS SECTION-A SECTION-B ;comment
> ```
> The section with the name `SECTION` is defined, the keyword `EXTENDS` refers
> to the derivation from the sections `SECTION-A` and `SECTION-B`. Thus,
> `SECTION` is based on the keys and values of the sections `SECTION-A` and
> `SECTION-B`. From the semicolon onwards, the following characters are
> interpreted as comments.

> __Line 2__
> ```
> PARAM-A = VALUE-1 ;comment
> ```
> The value `VALUE-1` is assigned to the key `PARAM-A`. The following characters
> are interpreted as comments from the semicolon onwards.

> __Line 3__
> ```
> PARAM-B [+] = VALUE-2; VALUE-3
> ```
> The key `PARAM-B` is assigned `VALUE-2; VALUE-3` as a value, with the option
> `[+]` at the end of the key disables the line comment and uses all characters
> for the value assignment, but it is not possible to specify a comment in this
> line.

> __Line 4__
> ```
> + VALUE-4; VALUE-5
> ``` 
> The value assignment of line 3 is continued and the value `VALUE-4;  VALUE-5`
> is added to the existing value of the key `PARAM-B`. The option `[+]` from
> line 3 is also taken from line 4, which also disables the line comment and
> uses all characters as value assignment. It is not possible to enter a comment
> in this line. Further options are not possible. 

> __Line 5__
> ```
> PARAM-C [?][+] = VALUE-6; VALUE-7
> ``` 
> The value assignment for the key `PARAM-C` is dynamic, and the system
> properties (VM arguments) of the Java runtime environment are searched for the
> key `PARAM-C` of the same name, ignoring upper and lower case. The key must be
> part of the runtime environment or can be set as VM argument (property) at the
> program start in the form `-Dkey=value`. If the system properties of the Java
> runtime environment do not contain a corresponding key, `VALUE-6; VALUE-7` is
> used as the value. By combining it with the `[+]` option at the end of the 
> key, the line comment is disabled and all characters are used for the value 
> assignment. It is not possible to enter a comment in this line. 

> __Line 6__
> ```
> PARAM-E [?] ;comment
> ``` 
> The value assignment for the key `PARAM-E` is dynamic, and the system 
> properties (VM arguments) of the Java runtime environment are searched for the
> key of the same name, ignoring upper and lower case. The key must be part of 
> the runtime environment or can be set as VM argument (property) at the program
> start in the form `-Dkey=value`. If the system properties of the Java runtime
> environment do not contain a corresponding key, this key is ignored because no
> alternative value has been specified. Comments are supported in this line.

> __Line 8 - 15__
> ```
> [0x53454354494F4E2D41]
>   PARAM-A                 = 0x574552542D31 ;comment
>   0x504152414D2D42        = VALUE-2        ;comment
>   0x504152414D2D43    [+] = 0x574552542D33
>   PARAM-D                 = 0x574552542D34
>                           + 0x574552542D35
>   PARAM-E          [?][+] = 0x574552542D363B20574552542D37
>   0x504152414D2D45 [?]                     ;comment
> ``` 
> Like the examples from lines 1 - 6, the hexadecimal notation is used for
> sections, keys and values.


## Configuration Structure

_Sections of the base configuration_
<table>
  <tr>
    <th>Section</th>
    <th>Description</th>
  </tr>
  <tr>
    <td><code>[COMMON]</code></td>
    <td>
      General application configuration.
    </td>
  </tr>
  <tr>
    <td><code>[INITIALIZE]</code></td>
    <td>
      Central module configuration. Initialization with the (re)start of the
      service.
    </td>
  </tr>
  <tr>
    <td><code>[REMOTE]</code></td>
    <td>
      Configuration area of the remote control of the server engine.
    </td>
  </tr>
  <tr>
    <td><code>[SERVER]</code></td>
    <td>
      Configuration area of the HTTP server.
    </td>
  </tr>
  <tr>
    <td><code>[VIRTUAL]</code></td>
    <td>
      Configuration area of the virtual hosts for HTTP servers.
    </td>
  </tr>
  <tr>
    <td><code>[STATUSCODES]</code></td>
    <td>
      HTTP server status codes.
    </td>
  </tr>
  <tr>
    <td><code>[MEDIATYPES]</code></td>
    <td>
      Assignment of media and data types.
    </td>
  </tr>
</table>

Further sections can be added as required.


## General Application Configuration
General and globally available options for server engine, servers, hosts and
modules are defined here.
        
_Overview of configuration_
<table>
  <tr>
    <th>Key</th>
    <th>Value</th>
    <th>Description</th>
  </tr>
  <tr>
    <td><code>CLEANUP</code></td>
    <td><code>ON|OFF</code></td>
    <td>
      Option for intelligent resource cleanup. The Java garbage collector is
      triggered if released resources are expected.
    </td>
  </tr>
  <tr>
    <td><code>RELOAD</code></td>
    <td><code>ON|OFF</code></td>
    <td>
      Option to monitor the configuration file and automatically reload and
      restart it when changes are made.
    </td>
  </tr>
</table>


## Initialisation
The section `[INITIALIZE]` is intended for global modules that are initiated
when the server engine is started and restarted.

> __Scheme for initialization__
> ```
> NAME = MODULE [OPTION] [OPTION] ...
> ```

> __Example of initialization__
> ```
> [INITIALIZE]
>   CONTROL = com.seanox.devwex.control.Connector
>   PHP     = com.seanox.fastcgi.Connector [connection:127.0.0.1:8129]
>   CGI     = com.seanox.fastcgi.Connector [connection:127.0.0.1:8139] [*]
> ```

With the option `[*]` a module can be marked as optional. If the module is not
present during initialization, this option does not cause to an error output.


## Server
Servers are an essential part of the server engine, and it requires at least one
successfully set up server instance to run. During initialization, all sections
that end with `INI` and for which a suitable implementation is searched for in
the class path are taken into account. The class name of the implementation is
the first part of the section name before the first colon.

> __Example of the structure__
> ```
> [CLASS:INI]
>   SCOPE = package.of.class
>   ...
>
> [NAME:INI]
>   SCOPE = package.of.class.Implementation
>   ...
>
> [CLASS:IDENTIFIER:INI]
>   SCOPE = package.of.class
>   ...
>
> [NAME:IDENTIFIER:INI]
>   SCOPE = package.of.class.Implementation
>   ...
> ```

Optionally, the key `SCOPE` can be used to specify the package in which the
class is located or the fully qualified class. Without this key, the default
package `com.seanox.devewx` is used and the class name is derived from the
section name. If no implementation is found for a section ending on `INI`, it
will be ignored without an error message.

> __Example of usage__
> ```
> [CLASS:INI]
>   SCOPE = package.of.class
>   ...
>
> [CLASS:IDENTIFIER:INI]
>   SCOPE = package.of.class
>   ...
> ```


## Remote Control Server
The included remote control supports a Telnet-based access of the server engine
(restart and stop) and queries about the operating status of the running servers
and modules. The server component of the remote control is a typical [servers](
    #server) configuration, except that it is an integral part of the
implementation and uses the fixed name `REMOTE` in the configuration.

> __Example of configuration__
> ```
> [REMOTE:INI]
>   ADDRESS = 127.0.0.1
>   PORT    = 25000
> ```

_Overview of configuration_
<table>
  <tr>
    <th>Key</th>
    <th>Value</th>
    <th>Description</th>
  </tr>
  <tr>
    <td><code>ADDRESS</code></td>
    <td><code>AUTO|IP|NAME</code></td>
    <td>
      Local address or name of the server in the network. <code>AUTO</code>
      corresponds to <code>0.0.0.0</code> and uses all IP addresses available in
      the system.
    </td>
  </tr>
  <tr>
    <td><code>PORT</code></td>
    <td><code>...</code></td>
    <td>
      Local port of the server.
    </td>
  </tr>
</table>


## HTTP / HTTPS Server
The server for hypertext transfer protocol has many configuration options, which
are divided into different sections, whose order is arbitrary. All the sections
have the same name as the server section, but differ in the section identifier
at the end.

_Overview of sections_
<table>
  <tr>
    <th>Section</th>
    <th>Description</th>
  </tr>
  <tr>
    <td><code>[SERVER:X:INI]</code></td>
    <td>
      Server configuration, network and runtime behavior
    </td>
  </tr>
  <tr>
    <td><code>[SERVER:X:SSL]</code></td>
    <td>TLS/SSL configuration</td>
  </tr>
  <tr>
    <td><code>[SERVER:X:REF]</code></td>
    <td>Virtual paths</td>
  </tr>
  <tr>
    <td><code>[SERVER:X:ACC]</code></td>
    <td>Access control (Access Control List)</td>
  </tr>
  <tr>
    <td><code>[SERVER:X:CGI]</code></td>
    <td>XCGI/CGI assignment and configuration</td>
  </tr>
  <tr>
    <td><code>[SERVER:X:ENV]</code></td>
    <td>Environment variables</td>
  </tr>
  <tr>
    <td><code>[SERVER:X:FLT]</code></td>
    <td>Filter configuration</td>
  </tr>
</table>


### Server
This section contains the configuration for HTTP servers and their associated
virtual hosts. The configurations for HTTP servers and virtual hosts is the
same, except for the network specific keys: `ADDRESS`, `PORT`, `BACKLOG` and
`MAXACCESS` which are omitted for virtual hosts, since they are configured only
server specific.

The server configuration consists of two parts: Network connection and HTTP. For

CHANGES
5.9.1 2027xxxx
??: Test: Docker runtime environment for LTS compatibility tests
??: Test: GitHub Action (scheduler) for performing LTS testing

5.9.0 20260411
NT: Celebrating 25 years of Devwex!
    Active development of new features has ended.
    Support and maintenance remain ongoing.
    Unit tests are run annually using the current Java LTS version.
    Corrections and minor requests will continue to be addressed.
BF: Review: Optimization and corrections
    And the binary becomes a bit smaller again :-)
BF: Windows: Correction in the determination of platform architecture
CR: HTTP(S) CGI: Optimization and corrections
    - Optimization in handling blocking CGI processes
    - Optimization in enforcing the end of the process
    - Optimization in the analysis of the exit code now consistently leads to
      status 502 in the access log and, if no header has been sent yet, also as
      HTTP status.
CR: HTTP(S): Optimization and corrections
    - Optimization in handling blocking I/O streams
    - Cleaning up old MIME types in the configuration
CR: HTTP(S) Directory Index: Optimization and corrections
    - Template: "case" and "type" are omitted without replacement
    - Template: "date" becomes "modified" and the sorting changes from d/D to m/M
    - Template: "sort" becomes "state", including the state logic in the CSS
    - Template: "mime" is now the fully qualified value
    - Generator value "-" for "no data" is omitted; now it's empty
CR: Review: Initialize refactored in Settings
CR: Review: Facelift status / index / templates
    - Use of system colors and thus support for light and dark mode
    - Improved display for HiDPI
CR: Windows: Omission from the 32-bit service
    - The required 32-bit version of prunsrv.exe can be downloaded manually
      here, renamed, and used to replace service.exe:
      https://downloads.apache.org/commons/daemon/binaries/windows/
CR: Documentation: Added installation as a service for Linux and macOS
CR: Configuration: Added support of # as alternative comment symbol
    The configuration now supports both ; and # as comment characters, which
    improves compatibility with common INI dialects and makes editing easier in
    modern editors.

5.8.0 20251231
NT: Competition: 25 years Devwex -- raffle one of 10x 3.5" disks.
    Send an e-mail to devwex@seanox.com by 04/11/2026.
    The outcome of the raffle is final and not subject to legal appeal. 
BF: Review: Optimization and corrections
BF: HTTP(S): Update / standardization of HTTP status messages
BF: HTTP(S): Omission of status 424, which was simply incorrect
    Status 400 and 500 are now used correctly.
CR: Test: Integration of JaCoCo (Java Code Coverage)
CR: Service: Optimization of logging output

5.7.0 20241214
BF: Unix/Linux: Correction of the missing second start parameter in the bash script
BF: Windows: Correction of the missing second start parameter in the batch script
BF: HTTP(S): Correction of the Accept header when using CGI
    The server has incorrectly determined the content type for the CGI.
    Now the CGI must take care of the Accept header, as it should.
BF: Service: Optimization of automatic resource clean-up
BF: Service: Optimization for UTF-8
    In relation to RFC 20, RFC 1345, RFC 2616, RFC 7230: 7-bit US-ASCII and in
    extension ISO-8859-1 are supported for request and response.
CR: Unix/Linux: Optimization of bash scripts and runtime directory handling
    - Optimization of JRE determination in the runtime directory
      The Directory name does not matter, it is searched for files ./java,
      ./jre/bin/java and ./bin/java.
    - Optimization when including the runtime sub-directories
      The runtime directories and, if available, the bin subdirectory are added
      to the PATH variable.
    - Change of the automatic runtime-script integration
      The ./runtime/scripts directory is omitted, the *.sh files are searched
      in the root of the ./runtime directory.
CR: Windows: Optimization of batch scripts and runtime directory handling
    - Optimization of JRE determination in the runtime directory
      The Directory name does not matter, it is searched for files .\java.exe,
      .\jre\bin\java.exe and .\bin\java.exe.
    - Optimization when including the runtime sub-directories
      The runtime directories and, if available, the bin subdirectory are added
      to the PATH variable.
    - Change of the automatic runtime-script integration
      The .\runtime\scripts directory is omitted, the *.bat and *.cmd files are
      searched in the root of the .\runtime directory.
CR: HTTP(S): Optimization of the composition of the UNIQUE_ID
CR: Documentation: Switch to Markdown
CR: Test: Review and Integration in build.xml (preparation for automation)
CR: Windows: Update of service-32.exe / service-64.exe (prunsrv.exe 1.4.0.0)
    https://dlcdn.apache.org//commons/daemon/binaries/windows/
    https://commons.apache.org/proper/commons-daemon/changes-report.html

5.6.0 20231227
BF: Review: Optimization and corrections
    And the binary becomes a bit smaller again :-)
BF: Worker: Correction of the unwanted SocketException on stop and restart
    The behavior is correct, only the error output is wrong.
BF: Worker: Correction of logging if the client closes the connection unexpectedly
    In some cases, the http_host variable for logging was not yet set.
BF: HTML: Cleanup of the obsolete type attribute in the style tag
BF: Project: Update of slogan (Experimental Server Engine)
BF: (Class)Loader: Optimization and corrections of error handling
    Important errors such as UnsupportedClassVersionError are no longer swallowed.
BR: Windows: Optimization and corrections for case insensitive commands
    concerning service.cmd
CR: XAPI: Change method String Module.explain() to String Module.expose()
CR: SAPI: Change method String Server.explain() to String Server.expose()
CR: Documentation: Reduction to English
CR: Test: Integration of the test environment

sources/com/seanox/devwex/Settings.java
/**
 * Devwex, Experimental Server Engine
 * Copyright (C) 2025 Seanox Software Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.seanox.devwex;

import java.math.BigInteger;
import java.util.Collections;
import java.util.Enumeration;
import java.util.LinkedHashMap;
import java.util.StringTokenizer;

/**
 * Settings, processes configuration data in INI format and provides it as
 * {@link Settings} and {@link Section}.<br>
 * <br>
 * The INI format used for configuration is a compatible. extension to the
 * classic format. It is also line-based and uses sections in which keys with
 * values are stored. Section and key names are case-insensitive. Declarations
 * with the same name cause sections and keys to be overwritten.<br>
 * <br>
 * Sections support multiple inheritance with the keyword {@code EXTENDS}. For
 * this purpose, after the keyword the names of the referenced sections are
 * specified from which the keys and values are to be applied. Thereby keys of
 * the same name of following sections overwrite already existing values.<br>
 * <br>
 * Assigning values to keys in a section is done with the equal sign. In
 * difference to the original format, the assignment can be continued in the
 * next line if the line starts with the plus sign.<br>
 * <br>
 * Values can be assigned final, variable and optional. The option {@code [?]}
 * at the end of a key is used to search for a key with the same name in the
 * system properties of the Java runtime environment. If none can be determined,
 * the optionally specified value is used. Without a value, such a key is not
 * specified and is ignored.<br>
 * <br>
 * Comments start with a semicolon, are optional, and can be used anywhere on a
 * line so that the following characters are not part of the section, key, or
 * value. The option {@code [+]} at the end of a key can be used to disable the
 * use of comments for that key and use the semicolon in the value.<br>
 * <br>
 * Sections, keys and values also support a hexadecimal notation, starting with
 * {@code 0x...} followed by the hexadecimal string, which can be used only
 * applied to the whole element.
 * 
 * <h3>Example</h3>
 * <pre>
 *   001 [SECTION] EXTENDS SECTION-A SECTION-B      ;comment
 *   002   PARAM-A                 = VALUE-1        ;comment
 *   003   PARAM-B             [+] = VALUE-2; VALUE-3
 *   004                           + VALUE-4; VALUE-5
 *   005   PARAM-C          [?][+] = VALUE-6; VALUE-7
 *   006   PARAM-E          [?]                     ;comment
 *   007
 *   008 [0x53454354494F4E2D41]
 *   009   PARAM-A                 = 0x574552542D31 ;comment
 *   010   0x504152414D2D42        = VALUE-2        ;comment 
 *   011   0x504152414D2D43    [+] = 0x574552542D33
 *   012   PARAM-D                 = 0x574552542D34
 *   013                           + 0x574552542D35
 *   014   PARAM-E          [?][+] = 0x574552542D363B20574552542D37
 *   015   0x504152414D2D45 [?]                     ;comment
 * </pre>
 * 
 * <h4>Line 1</h4>
 * The section with the name {@code SECTION} is defined, the keyword
 * {@code EXTENDS} refers to the derivation from the sections {@code SECTION-A}
 * and {@code SECTION-B}. Thus, {@code SECTION} is based on the keys and values
 * of the sections {@code SECTION-A} and {@code SECTION-B}. From the semicolon
 * onwards, the following characters are interpreted as comments.
 * 
 * <h4>Line 2</h4>
 * The value {@code VALUE-1} is assigned to the key {@code PARAM-A}. The
 * following characters are interpreted as comments from the semicolon onwards.
 * 
 * <h4>Line 3</h4>
 * The key {@code PARAM-B} is assigned {@code VALUE-2; VALUE-3} as a value, with
 * the option {@code [+]} at the end of the key disables the line comment and
 * uses all characters for the value assignment, but it is not possible to
 * specify a comment in this line.
 * 
 * <h4>Line 4</h4>
 * The value assignment of <i>line 3</i> is continued and the value {@code
 * VALUE-4; VALUE-5} is added to the existing value of the key {@code PARAM-B}.
 * The option {@code [+]} from <i>line 3</i> is also taken from <i>line 4</i>,
 * which also disables the line comment and uses all characters as value
 * assignment. It is not possible to enter a comment in this line. Further
 * options are not possible.
 * 
 * <h4>Line 5</h4>
 * The assignment for the key {@code PARAM-C} is dynamic, and the system
 * properties (VM arguments) of the Java runtime environment are searched for
 * key {@code PARAM-C} of the same name, case-insensitive. The key must be part
 * of the runtime environment or can be set as VM argument (property) at the
 * program start in the form {@code -Dkey=value}. If the system properties of
 * the Java runtime environment do not contain a corresponding key, {@code
 * VALUE-6; VALUE-7} is used as the value. By combining it with the option
 * {@code [+]} at the end of the key, the line comment is disabled and all
 * characters are used for the value assignment. It is not possible to enter a
 * comment in this line.
 * 
 * <h4>Line 6</h4>
 * The value assignment for the key {@code PARAM-E} is dynamic, and the system
 * properties (VM arguments) of the Java runtime environment are searched for
 * the key of the same name, case-insensitive. The key must be part of the
 * runtime environment or can be set as VM argument (property) at the program
 * start in the form {@code -Dkey=value}. If the system properties of the Java
 * runtime environment do not contain a corresponding key, this key is ignored
 * because no alternative value has been specified. Comments are supported in
 * this line.
 * 
 * <h4>Line 8 - 15</h4>
 * Like the examples from <i>lines 1 - 6</i>, the hexadecimal notation is used
 * for sections, keys and values.
 */
public class Settings implements Cloneable {

    /** Map of sections */
    private final LinkedHashMap entries;
    
    /** Option to activate smart mode */
    private final boolean smart;

    /** Constructor, creates Settings. */
    public Settings() {
        this(false);
    }
    
    /** 
     * Constructor, creates Settings.
     * Optionally, smart behavior can be enabled, which changes the behavior of
     * some methods.
     * @param smart activates smart mode
     */
    public Settings(boolean smart) {
        this.entries = new LinkedHashMap();
        this.smart   = smart;
    }

    /**
     * Decodes hexadecimal values into a string if necessary.
     * @param  string String to decode
     * @return the decoded and trimmed string
     */                   
    private static String decode(String string) {
        if (string == null)
            string = "";
        string = string.toUpperCase().trim();
        if (string.matches("^(?i)0x([0-9a-f]{2})+$"))
            return new String(new BigInteger(string.substring(2), 16).toByteArray()).toUpperCase().trim();
        return string;
    }

    /**
     * Determines the contained sections from the string.
     * Parsing ignores invalid sections, keys and values and returns the
     * determined sections as Settings.
     * @param  text String to parse
     * @return the determined sections as Settings
     */
    public static Settings parse(String text) {
        return Settings.parse(text, false);
    }
    
    /**
     * Determines the contained sections from the string.
     * Parsing ignores invalid sections, keys and values and returns the
     * determined sections as Settings.
     * 
     * Optionally, smart behavior can be enabled, which changes the behavior of
     * some methods.
     *     
     *     <dir>{@link #parse(String, boolean)}</dir>
     * Only sections that are not empty are applied.
     *     
     *     <dir>{@link #get(String)}</dir>
     * Never returns {@code null} for valid keys. If an unknown section is
     * requested, it will be recreated.
     *     
     *     <dir>{@link #set(String, Section)}</dir>
     * If {@code null} is passed as the value of a section key, the method
     * behaves like {@link #remove(String)} and deletes the section.    
     * 
     *     <dir>{@link #merge(Settings)}</dir>
     * Only sections that are not empty are applied. If a section contain value
     * {@code null}, the method behaves like {@link #remove(String)}.
     * 
     * @param  text  String to parse
     * @param  smart Activates smart mode
     * @return the determined sections as Settings
     */
    public static Settings parse(String text, boolean smart) {

        Settings settings = new Settings(smart);

        if (text == null)
            return settings;

        LinkedHashMap entries = new LinkedHashMap();
        StringBuffer  buffer  = null;

        StringTokenizer tokenizer = new StringTokenizer(text, "\r\n");
        while (tokenizer.hasMoreTokens()) {

            String line = ((String)tokenizer.nextElement()).trim();
            if (line.startsWith("[")) {
                
                buffer = new StringBuffer();
                
                // The line is processed as follows:
                //   - the unique/valid name of the section is determined
                //   - the comment part is removed
                //   - inadmissible subsequent sections are removed
                //   - if necessary existing derivations are loaded
                String[] strings = line.replaceAll("^(?i)(?:\\[\\s*([^\\[\\]\\#;]+)\\s*\\]\\s*(?:extends\\s+([^\\[\\]\\#;]+))*)*.*$", "$1 \00 $2").split("\00");
                
                // The section will be decoded and optimized if necessary but
                // only valid sections will be loaded
                String section = Settings.decode(strings[0]);
                if (section.length() <= 0)
                    continue;
                entries.put(section, buffer);
                
                // Any existing derivations are registered and loaded
                strings = strings[1].split("\\s+");
                for (int index = 0; index < strings.length; index++) {
                    section = Settings.decode(strings[index]);
                    if (entries.containsKey(section))
                        buffer.append("\r\n").append(entries.get(section));
                }
                
            } else if (buffer != null) {

                // Content is accepted only for valid section
                buffer.append("\r\n").append(line);
            }
        }
        
        Enumeration enumeration = Collections.enumeration(entries.keySet());
        while (enumeration.hasMoreElements()) {
            String section = (String)enumeration.nextElement();
            String value   = ((StringBuffer)entries.get(section)).toString().trim();
            if (value.length() > 0
                    || !smart)
                settings.entries.put(section, Section.parse(value, smart));
        }

        return settings;
    }

    /**
     * Returns an enumeration of the keys of all sections.
     * @return the keys of all sections as enumeration
     */
    public synchronized Enumeration elements() {
        return Collections.enumeration(this.entries.keySet());
    }

    /**
     * Returns {@code true} if the key to a section is contained.
     * @param  key Section name
     * @return {@code true} if the key to a section is contained
     */
    public synchronized boolean contains(String key) {
        if (key != null)
            key = key.toUpperCase().trim();
        if (key == null
                || key.length() <= 0)
            return false;
        return this.entries.containsKey(key);
    }

    /**
     * Returns the section corresponding to the key. If the key is not contained
     * or invalid, {@code null} is returned. In smart mode, never returns
     * {@code null} for valid keys. If an unknown section is requested, it will
     * be recreated.
     * @param  key Section name
     * @return the determined section, otherwise {@code null} 
     */
    public synchronized Section get(String key) {

        if (key != null)
            key = key.toUpperCase().trim();
        if (key == null
                || key.length() <= 0)
            return null;
        
        Section section = (Section)this.entries.get(key);
        if (section == null
                && this.smart) {
            section = new Section(true);
            this.set(key, section);
        }
        return section;
    }

    /**
     * Sets the corresponding section with the specified content. Existing
     * sections are replaced. Empty keys and {@code null} are ignored and return
     * {@code null}. In smart mode sections are removed if {@code null} is
     * passed as value. The behavior is the same as for {@link #remove(String)}. 
     * @param  key     Section name
     * @param  section Section
     * @return previously assigned section, otherwise {@code null}
     */
    public synchronized Section set(String key, Section section) {

        if (key == null)
            return null;
        key = key.toUpperCase().trim();
        if (key.length() <= 0)
            return null;
        
        if (section == null
                && !this.smart)
            return (Section)this.entries.remove(key);
        if (section == null)
            section = new Section(this.smart);
        return (Section)this.entries.put(key, section);
    }

    /**
     * Removes the specified section from Settings.
     * @param  key Name of the section to be removed
     * @return previously assigned section, otherwise {@code null}
     */
    public synchronized Section remove(String key) {
        if (key != null)
            key = key.toUpperCase().trim();
        if (key == null
                || key.length() <= 0)
            return null;
        return (Section)this.entries.remove(key);
    }

    /**
     * Merges the passed sections. Existing sections are updated and new
     * sections are created. In smart mode, only sections that are not empty are
     * applied. If a section contain value {@code null}, the method behaves like
     * {@link #remove(String)}.
     * @param  settings Sections to be applied
     * @return the current instance with the merged sections
     */
    public synchronized Settings merge(Settings settings) {
        
        if (settings == null)
            return this;

        // Sections are combined or, if necessary, newly created
        Enumeration enumeration = Collections.enumeration(this.entries.keySet());
        while (enumeration.hasMoreElements()) {
            String entry = (String)enumeration.nextElement();
            Section section = settings.get(entry);
            this.set(entry, section.merge(this.get(entry)));
        }
        
        return this;
    }

    /**
     * Returns the number of sections.
     * @return number of sections
     */
    public synchronized int size() {
        return this.entries.size();
    }

    /** Resets Settings completely and discards all sections. */
    public synchronized void clear() {
        this.entries.clear();
    }

    @Override
    public synchronized Object clone() {
        Settings settings = new Settings(this.smart);
        Enumeration enumeration = Collections.enumeration(this.entries.keySet());
        while (enumeration.hasMoreElements()) {
            String entry = (String)enumeration.nextElement();
            settings.entries.put(entry, ((Section)this.entries.get(entry)).clone());
        }
        return settings;
    }
}